### PR TITLE
Introduce POSIX Signal API

### PR DESCRIPTION
**Pull request by JeroenSoeters**
This PR will introduce a method to the `Observe` API that returns a stream of all [signals](https://man7.org/linux/man-pages/man7/signal.7.html) _handled_ on the host.

### CODE REVIEW

File: auraed/src/observe/mod.rs
Comment (dmah42) [inline auraed/src/observe/mod.rs:147]: https://docs.rs/signal/0.5.1/signal/enum.Signal.html ?

File: Makefile
Comment (dmah42) [inline Makefile:286]: i'm not sure the alias is adding much here.. explicit is nice :)

is it useful for local dev to have it? i assume not if we can `make ebpf`.

File: aurae-ebpf-shared/src/lib.rs
Comment (dmah42) [inline aurae-ebpf-shared/src/lib.rs:6]: should we use a predefined enum here? https://docs.rs/signal/0.5.1/signal/enum.Signal.html exists and i assume it supports no_std...

File: Makefile
Comment (JeroenSoeters) [reply Makefile:286]: I started with `xtask` but can probably get rid of it now it's in the `Makefile`. Will have to clean this up anyways as we should probably have the `auraed` build depend on the `ebpf` target.

File: aurae-ebpf-shared/src/lib.rs
Comment (JeroenSoeters) [reply aurae-ebpf-shared/src/lib.rs:6]: Unfortunately we cannot use that from the kernel. The `Signal` crate depends on `nix` which pulls in a whole slew of transitive dependencies which cause the `aurae-ebpf` build to fail. We could have `load_tracepoint<T>` or the `PerfEventListener<T>` take an extra type parameter, which could be a more idiomatic `Rust` type that would implement `From` for the C-struct that has to traverse the FFI boundary but then again in this codepath we'd be immediately mapping it again to a `proto` type.

File: aurae-ebpf/Cargo.toml
Comment (future-highway) [inline aurae-ebpf/Cargo.toml:2]: Would we only every have a single ebpf program, or would we have 1 per task. If it is the latter, then the crate name should be more specific.

File: auraed/src/logging/log_channel.rs
Comment (future-highway) [inline auraed/src/logging/log_channel.rs:41]: Is this struct being used outside the crate?

File: auraed/src/lib.rs
Comment (future-highway) [inline auraed/src/lib.rs:304]: Is the `Arc` needed? `LogChannel` can be cloned.

File: xtask/src/build_ebpf.rs
Comment (JeroenSoeters) [reply xtask/src/build_ebpf.rs:60]: I'm getting rid of `xtask` altogether. I just started off with an `aya` template and moved that over to our codebase. I don't see any value keeping `xtask`, and as long as we're targeting `x86_64-*` we'll always have the little endian target (`bpfel-*`).

File: aurae-ebpf/Cargo.toml
Comment (JeroenSoeters) [reply aurae-ebpf/Cargo.toml:2]: So the way I was thinking about it is having a single binary that would include all the eBPF programs. The programs are named via this [aya attribute](https://github.com/JeroenSoeters/aurae/blob/posix-signals-api/aurae-ebpf/src/main.rs#L17). The `BpfLoader` will then [load these programs](https://github.com/JeroenSoeters/aurae/blob/posix-signals-api/auraed/src/ebpf/loader.rs#L42) from the elf file `aurae-ebpf` into the kernel and [attach them to the correct trace event](https://github.com/JeroenSoeters/aurae/blob/posix-signals-api/auraed/src/ebpf/loader.rs#L48).  I will rename this `name` parameter on `load_tracepoint` to `prog_name` to make it more explicit that it's referring to an actual eBPF program instead of being some random name.

File: xtask/src/build_ebpf.rs
Comment (JeroenSoeters) [reply xtask/src/build_ebpf.rs:60]: Should we just build it in the root target folder, btw? Not sure what's idiomatic here since it is not a member of the workspace.

File: auraed/src/lib.rs
Comment (JeroenSoeters) [reply auraed/src/lib.rs:304]: I don't think that `LogChannel` is actually used. I just did the minimum to make it compile 😅

File: xtask/src/build_ebpf.rs
Comment (future-highway) [reply xtask/src/build_ebpf.rs:60]: What is the reasoning for not having it as part of the workspace?

Comment (JeroenSoeters) [reply xtask/src/build_ebpf.rs:60]: I'm not sure why but a lot of extra stuff was ending up in the binary when I included it in the workspace. I think it was like 10-20x times the size and `aya` wasn't able to perform the relocations on the `elf` file (the binary). I tried a few things but then I gave up as excluding it was easier. :) If you want to take a look and see what's going on, you could compile it as a workspace member and inspect the output file with `objdump -S <filename>`. All the names are mangled though so it can be a little hard on the eyes.

Comment (future-highway) [reply xtask/src/build_ebpf.rs]: If it can't be part of the workspace that is fine. It should have its own target directory.

Comment (JeroenSoeters) [reply xtask/src/build_ebpf.rs]: Cool, `xtask` is gone now and the `Makefile` updated. Note that we do have a dependency on `nightly` now just for this project only as well as a dependency on `bpf-linker`. Once this PR is in I'll update the docs in the `aurae.io` repo.

File: auraed/src/logging/log_channel.rs
Comment (JeroenSoeters) [reply auraed/src/logging/log_channel.rs:41]: We wouldn't be able to expose it in `ObserverService::new` if it is crate-private that's why I made the change. Before the `ObserveService` wasn't actually used.

**Review by dmah42 — CHANGES_REQUESTED** [review]
putting this here as a merge blocker until after the weekend (live demo for FOSDEM)

**Review by dmah42 — COMMENTED** [review]
unblocking now the talk is done.

File: auraed/Cargo.toml
Comment (future-highway) [inline auraed/Cargo.toml:72]: Is this dependency being used? There is a newer version of it also, is there a reason to use the older version? If we only want a signal enum, nix has one at `nix::sys::signal::Signal`

File: auraed/src/lib.rs
Comment (future-highway) [inline auraed/src/lib.rs:281]: I'd prefer is we had a fn on `BpfLoader` named `load_signals_tracepoint`

File: auraed/src/observe/mod.rs
Comment (future-highway) [inline auraed/src/observe/mod.rs:55]: We can remove this.

File: auraescript/src/lib.rs
Comment (future-highway) [inline auraescript/src/lib.rs:97]: Nit pick: can we make this alphabetical too (cri is also in the wrong spot)

File: docs/build.md
Comment (future-highway) [inline docs/build.md:28]: I have concerns about licensing issues and want to make sure we get a review from @krisnova to see if this is ok to add before we merge anything.

**Review by future-highway — CHANGES_REQUESTED** [review]
Blocking until licensing concerns are alleviated.

In my very short amount of research, it seems like eBPF programs declare their license and the kernel will only expose things that are compatible with such license. Diving into aya (which aurae-ebpf) is build on), it seems like it trys to parse a license and if it fails to parse, the [default is gpl](https://github.com/aya-rs/aya/blob/7868fffd7f3fd76611d87d5a0385dbd59ad72f48/aya-obj/src/obj.rs#L488). That gives me concerns. Is there a way to test that the correct license is being parsed?

There is also some guidance on licensing [in the kernel docs](https://docs.kernel.org/bpf/bpf_licensing.html).

File: docs/build.md
Comment (dmah42) [reply docs/build.md:28]: not trying to resolve as that's up to @krisnova, but the use of GPL for eBPF is a well-known thing and other projects that embed eBPF programs end up dual-licensed which is _fine_ but a bit of a pain.

can we just pass in a license?

File: auraed/src/lib.rs
Comment (JeroenSoeters) [reply auraed/src/lib.rs:281]: I was hoping to keep the `BpfLoader` agnostic of which actual programs are loaded and just supply it with a couple of methods like `load_tracepoint`, `load_kprobe`, `load_uprobe` etc and maintain a list of the actual programs we want to load elsewhere. I can add an extra function for now until we figure out where this list would live.

File: auraed/src/logging/log_channel.rs
Comment (future-highway) [reply auraed/src/logging/log_channel.rs:41]: We'll refactor in a future PR

File: aurae-ebpf-shared/Cargo.toml
Comment (future-highway) [reply aurae-ebpf-shared/Cargo.toml:5]: Typo "publish", not "package"

File: xtask/src/build_ebpf.rs
Comment (future-highway) [reply xtask/src/build_ebpf.rs]: This repo is also used for aurae.io. The other repo is archived.

File: auraed/src/lib.rs
Comment (future-highway) [reply auraed/src/lib.rs]: I understand your reasoning and would agree if this were library code, but since this is internal I think the helper methods are the right approach. Thanks for making the change.

File: docs/build.md
Comment (JeroenSoeters) [reply docs/build.md:28]: The licensing pertains to BPF helper functions in the kernel, [some of them are GPL-ed and some are not](https://github.com/iovisor/bcc/blob/master/docs/kernel-versions.md#helpers). This PR is using [`BPF_FUNC_perf_event_output()`](https://github.com/JeroenSoeters/aurae/blob/posix-signals-api/aurae-ebpf/src/main.rs#L44) which has a GPL license. I think we would need to dual license only the BPF code [like Cilium does](https://github.com/cilium/cilium#license).

File: xtask/src/build_ebpf.rs
Comment (JeroenSoeters) [reply xtask/src/build_ebpf.rs]: Yep found it already and included the changes in this PR.

File: docs/build.md
Comment (future-highway) [reply docs/build.md:28]: If the function we are using is GPL-ed and everything was working (which I assume is true given the discord post), then **`aya` must be assuming we are licensed under GPL**.

Some food-for-thought:
- If some code is going to be GPL-ed, should we put it in a separate repo with its own builds and releases? Even if it doesn't make a technical difference, I think that is a better approach then what I assume is the alternative -- update the license files to say "this is the default unless otherwise indicated".
- Do we need to update the CLA?

Also, I want to clarify that I'm not trying to suggest the direction we go in; I don't have the knowledge needed to make those suggestions.

File: auraescript/src/lib.rs
Comment (future-highway) [inline auraescript/src/lib.rs:71]: Nitpick: alphabetical

Sidenote: I think rustfmt would take care of this automatically, so you may not have it turned on.

Also, don't think of the nitpicks as blockers. I wouldn't block the merge just for those.

File: auraed/src/lib.rs
Comment (future-highway) [inline auraed/src/lib.rs:280]: This seems like it loads the bpf program in all contexts (PID 1, nested, container, etc.). Is that correct?

@krisnova or @dmah42

File: docs/build.md
Comment (JeroenSoeters) [reply docs/build.md:28]: > aya must be assuming we are licensed under GPL.

It's not just `aya`, you literally have to tell the kernel that the BPF program you're loading is GPL licensed. This is what that `syscall` looks like:

```bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_TRACEPOINT, insn_cnt=31, insns=0x5555593ae5c0, license="GPL", log_level=0, log_size=0, log_buf=NULL, kern_version=KERNEL_VERSION(6, 1, 8), prog_flags=0, prog_name="signals", prog_ifindex=0, expected_attach_type=BPF_CGROUP_INET_INGRESS, prog_btf_fd=0, func_info_rec_size=0, func_info=NULL, func_info_cnt=0, line_info_rec_size=0, line_info=NULL, line_info_cnt=0, attach_btf_id=0, attach_prog_fd=0, fd_array=NULL}, 144)```

I'm not sure if we would want to create a separate repo for the BPF code, though. We'd have to run a private crate registry, and we'd be introducing a lot of friction in the development process, not to mention that for every change, we would now need 2 PRs instead of 1.

If you look at some of the other popular OSS eBPF projects like Cilium and Falco, they typically have a mention somewhere of which files are dual licensed and have a different license header for those files.

https://github.com/cilium/cilium#license
https://github.com/falcosecurity/libs/blob/master/NOTICES

File: auraed/src/lib.rs
Comment (JeroenSoeters) [reply auraed/src/lib.rs:280]: Yea I was thinking about this as well, I am going to change this to only load the BPF programs in the host daemon for now. I'm actually not fully understanding the `observe` gRPC endpoint tbh. Is there a way to `observe` a nested `auraed` right now? If there is we'd anyways need to think about mapping the host `PID` to the namespaced `PID` in cases where the `PID namespace` is `unshared`.

Comment (JeroenSoeters) [reply auraed/src/lib.rs]: Just changed this. eBPF is only loaded from the host daemon for now.

Comment (JeroenSoeters) [reply auraed/src/lib.rs]: Left some comments [on this issue](https://github.com/aurae-runtime/aurae/issues/366) on how we could `observe` signals for nested auraed processes. Probably should do this in a different PR though.

File: Makefile
Comment (dmah42) [inline Makefile:241]: is this different to `$(cargo) +nightly build --path ./aurae-ebpf --package ...` ?

File: docs/build.md
Comment (dmah42) [reply docs/build.md:28]: dual-licensing may not be an issue for our use-case as we're not shipping this as a library to be integrated into other projects.  ie, anyone _using_ aurae is not going to be subject to the terms in GPL unless they're building code linking aurae in, which is not what i'd expect.

but IANAL and i agree @krisnova should have final say.  also, i don't feel qualified to assess the eBPF parts of this PR :)

File: Makefile
Comment (JeroenSoeters) [reply Makefile:241]: I'm not showing `--path` as a valid `cargo build` flag. I'm on cargo 1.66.1.

```error: unexpected argument '--path' found```

File: auraescript/src/lib.rs
Comment (JeroenSoeters) [reply auraescript/src/lib.rs:68]: I do have `rustfmt` enabled on save so I'm a bit puzzled as well as why it doesn't order these automatically.

File: Makefile
Comment (dmah42) [reply Makefile:241]: ah looks like we use install for other executables.

**Comment by krisnova** [comment]
Hey everyone jumping in I am going to look at the code now and possibly make some superficial changes to get this merged in!

**Comment by krisnova** [comment]
Also we will need all contributors to this PR to add a GitHub comment with the following language:

> I hereby agree to the terms and conditions in the GNU Public License (GPL) Version 2 for all work contributed.

**Comment by JeroenSoeters** [comment]
I hereby agree to the terms and conditions in the GNU Public License (GPL) Version 2 for all work contributed.

**Review by krisnova — APPROVED** [review]

**Pull request merged**
